c++/src/kj/compat/http.c++

```diff
@@ -2637,6 +2637,31 @@ public:
 
   uint64_t receivedByteCount() override { return receivedBytes; }
 
+  kj::Maybe<kj::String> getPreferredExtensions(ExtensionsContext ctx) override {
+    if ((maskKeyGenerator == nullptr && ctx == ExtensionsContext::REQUEST) ||
+        (maskKeyGenerator != nullptr && ctx == ExtensionsContext::RESPONSE)) {
```

Comment:
I think we're confusing the meaning of REQUEST and RESPONSE here.

If `maskKeyGenerator` is non-null, then this WebSocket object is the client side of a connection (since only the client needs to apply a mask to outgoing messages).

Meanwhile, if the caller passed `RESPONSE`, that means they are requesting to construct an extension header for use in a response, i.e. to create the server side of a connection.

That's the case we want to support -- coupling a client-side and a server-side WebSocket. But it looks like the code here will return null in that case?

Or am I reading this wrong?

Reply:
Right, sorry, got mixed up thinking about what the caller wants vs. what we want to extract from `this` websocket. It felt weird that `ctx` being `REQUEST` would result in us calling `generateExtensionResponse()`.

Just for posterity, what we want is:
**Caller**: "I want to pump from `this` to `other`, where `other` will be returned in a `RESPONSE`.
**Callee**:
1. Make sure that `this` is a client (non-null `maskKeyGenerator`), as `other` is a server-side connection.
2. Get `this` websocket's extensions. Since we're a client, `out->client`, so use `generateExtensionRequest`.
3. Return `permessage-deflate; ...`

**Caller**: Interprets `permessage-deflate; ...` parameters in the opposite direction, i.e. `out->server`, so we can `optimizedPump`.

Reply:
> It felt weird that ctx being REQUEST would result in us calling generateExtensionResponse().

Hmm, is that right? `generateExtensionResponse()` formats a response header, right? So that's what we should use if the client requests RESPONSE?

I think it was just this check that was backwards here, but the code below was correct originally (aside from the comment being confusing)?

Reply:
> `generateExtensionResponse()` formats a response header, right?

Yes, but more specifically it assumes that the `config` passed to it belongs to a server-side websocket. It assumes this because we map `inbound` parameters to `client`, and `outbound` to `server`.

> So that's what we should use if the client requests RESPONSE?

Bit unclear here, is `this` websocket a "client", i.e. we want to pump to a server, or did you mean "caller" in place of "client"?

---
TL;DR of the stuff below is: when the caller passes `RESPONSE` to `getPreferredExtensions()`, is the caller saying: 
```
a) I intend to pump from `this` websocket to a server-side websocket (implies `this` must be a client), or
b) I want `this` websocket's config to be parsed as though `this` websocket is server-side.
```
---
(This is a bit long, but I want to be clear so it's easier to find where the confusion is).
I think the main goal here is that we have some websocket, and we want to convert its `inboundX` parameters to `clientX` or `serverX`. Let's say we have a client-side websocket `proxyClient`, which has the following compression config:
```
CompressionParameters {
    outboundNoContextTakeover = false,
    inboundNoContextTakeover = true,
    outboundMaxWindowBits = 10,
    inboundMaxWindowBits = 13
};
```
We would want this to be converted to: 
```
permessage-deflate; server_no_context_takeover; client_max_window_bits=10; server_max_window_bits=13
```
since `outbound->client` and `inbound->server` for client-side websockets.

Now let's say we want to have another websocket, `proxyServer`, which is a server-side websocket. We want to have the final connection look like this:
```
eyeball <-----> proxyServer <==optimizedPumpTo==> proxyClient <-----> server

     outboundNoContextTakeover = true          outboundNoContextTakeover = false
     inboundNoContextTakeover = false          inboundNoContextTakeover = true
     outboundMaxWindowBits = 13                outboundMaxWindowBits = 10
     inboundMaxWindowBits = 10                 inboundMaxWindowBits = 13
```

To enable `optimizedPumpTo`, we need to ensure `proxyServer` uses extension parameters that are compatible with `proxyClient`. We would do this by passing the exact header value `permessage-deflate; server_no_context_takeover; client_max_window_bits=10; server_max_window_bits=13` to `acceptWebSocket`, because `acceptWebSocket` will map `server->outbound` and `client->inbound` for us.

**This is probably where the confusion is**:
As I see the _current_ code, calling `proxyClient.getPreferredExtensions(RESPONSE)` is the same as saying
> I want to convert `proxyClient`s compression config to a `Sec-WebSocket-Extensions` header, and I will use this header to create a websocket (`proxyServer`) that will act as a server-side websocket (used in a `RESPONSE`) that can `optimizedPumpTo` `proxyClient`.

This is based on what you said above:
> Meanwhile, if the caller passed `RESPONSE`, that means they are requesting to construct an extension header for use in a response, i.e. to create the server side of a connection.

Did you mean passing `RESPONSE` tells `getPreferredExtensions` to generate extensions as though `this` websocket is server-side?

So all things considered, `proxyClient.getPreferredExtensions(RESPONSE)` results in us converting `proxyClient`'s `CompressionParameters` to `permessage-deflate; server_no_context_takeover; client_max_window_bits=10; server_max_window_bits=13` (this is done by calling `generateExtensionRequest`, since `proxyClient` is client-side). We then pass this header value to `acceptWebSocket()`, which gives us `proxyServer`, which has the following `CompressionParameters`:
```
CompressionParameters {
    outboundNoContextTakeover = true,
    inboundNoContextTakeover = false,
    outboundMaxWindowBits = 13,
    inboundMaxWindowBits = 10
};
```
Which is the opposite of `proxyClient` (which is what we want).

Reply:
> Yes, but more specifically it assumes that the config passed to it belongs to a server-side websocket. It assumes this because we map inbound parameters to client, and outbound to server.

Ohhhhhh! I get it.